fix(animations): Trigger leave animation when ViewContainerRef is injected

JeanMeche opened the pull request:
Injecting ViewContainerRef into a component makes it effectively a container. The leave animation was triggered on container
    
fixes #48667

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix


## Does this PR introduce a breaking change?

- [x] No

JeanMeche commented:
I will remove the now unused `isHostComponent` argument if the tests are successful.

jessicajaniuk commented:
This will very likely need a TGP to verify it doesn't cause any breaking behaviors.

jessicajaniuk commented:
[TGP](https://fusion2.corp.google.com/presubmit/tap/507645240/OCL:507645240:BASE:507644627:1675739717977:d57f5616/targets)

jessicajaniuk commented:
The TGP looks green. So I'll take a look at this code today.

jessicajaniuk reviewed (changes requested):
The code itself looks ok, but it's unclear to me what case this is fixing. Can you add a specific test to cover this so we have clarity on what's broken, what's being fixed, and why this change is needed?

JeanMeche commented:
I wanted to add a test to `animation_integration_spec.ts` I didn't find a way to run it. Any idea ? 😊

jessicajaniuk commented:
@JeanMeche those tests are part of `packages/core/test`. So you can isolate and run them with `yarn bazel test packages/core/test`. The trick is that line 22 of `animations_integration_spec.ts` has `if (isNode) return;`, which skips them when running locally. So I typically just comment that out and focus the test I want.

JeanMeche commented:
Thanks for the heads-up Jessica, you pointed me into the right direction, I needed to run the test case in karma with ``yarn test packages/core/test:test_web`

The integration test I added was broken before the fix but passes fine with the fix.

jessicajaniuk reviewed (approved):
Thanks for adding the test.

reviewed-for: fw-core, fw-animations, public-api

pkozlowski-opensource commented:
> I will remove the now unused isHostElement argument if the tests are successful.

@JeanMeche so it seems like this PR has 2 changes:
* the actual fix in packages/animations/browser/src/render/transition_animation_engine.ts
* public API change of the Renderer2

While we can land a fix any time, changing API surface would mean going through the standard depreciation policy and the actual removal can happen only in a major version.

It seems to me like we can't just remove an argument of the Renderer2 since the existing code could be still calling renderer2 with the removed argument, right?

If the above makes sense we should split this PR into a fix _and_ a cleanup. WDYT?

JeanMeche committed:
fix(animations): Trigger leave animation when ViewContainerRef is injected

Injecting `ViewContainerRef` into a component makes it effectively a container. The leave animation wasn't triggered on containers before this fix.

fixes angular#48667

JeanMeche commented:
Thank you Pawel, you're right. I was still a beginner when I wrote that fix 😊

We'll do the deprecation in another PR.

pkozlowski-opensource commented:
Thnx @JeanMeche !

AndrewKushnir commented:
@jessicajaniuk since the last TGP was a while ago, could you please help run a new one (just to make sure there are no regressed targets in g3)? Thank you.

jessicajaniuk commented:
[New TGP](https://fusion2.corp.google.com/presubmit/507645240/OCL:507645240:BASE:534814923:1684935665446:d2191a9b/targets)

dylhunn commented:
This PR was merged into the repository by commit bada9199f53cb631d065a961427a3a1a8d9762f1.
